Donation-platform/src/main/java/com/google/servlets/UserNameServlet.java

```diff
@@ -0,0 +1,103 @@
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.KeyFactory.Builder;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import java.util.Collection;
+import java.util.Collections;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.security.GeneralSecurityException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/* Servlet that returns user information when logged in.*/
+@WebServlet("/username")
+public class UserNameServlet extends HttpServlet
+{
+
+    private static final String CLIENT_ID = "223187457231-nspsjgjtsnpgjub4q12p37cdu134d6kk.apps.googleusercontent.com";
+
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException
+    {
+
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        DbCalls dbCalls = new DbCalls(datastore);
+
+        GoogleIdTokenVerifier verifier = new GoogleIdTokenVerifier.Builder(new NetHttpTransport(), new JacksonFactory())
+        .setAudience(Collections.singletonList(CLIENT_ID))
+        .build();
+
+        try
+        {
+            GoogleIdToken idToken = verifier.verify(request.getParameter("idtoken"));
+            if (idToken != null)
+            {
+                // Returns profile information of current user.
+                Payload payload = idToken.getPayload();
+                Entity entity;
+                try
+                {
+                    // Reads user information from datastore.
+                    entity = datastore.get(KeyFactory.createKey("Users", payload.getSubject()));
+                }
+                // If user not in datastore, add user to datastore.
+                catch (EntityNotFoundException e)
+                {
+                    dbCalls.addUser(payload.getSubject(),(String) payload.get("name"), payload.getEmail(), Collections.emptyList(), Collections.emptyList());
+                    entity = datastore.get(KeyFactory.createKey("Users", payload.getSubject()));
+                }
+                Gson gson = new Gson();
+                String json = gson.toJson(dbCalls.setUsersClass(entity));
+                response.setContentType("application/json;");
+                response.getWriter().println(json);
+
+            }
+            else
+            {
+                System.out.println("Invalid ID token.");
```

Comment:
is this for debugging purpose/

Reply:
yes for us to see in the client side console.